src/librustc/ty/sty.rs

```diff
@@ -1809,51 +1809,61 @@ impl<'tcx> Const<'tcx> {
     }
 
     #[inline]
-    pub fn from_primval(
+    pub fn from_scalar(
         tcx: TyCtxt<'_, '_, 'tcx>,
-        val: PrimVal,
+        val: Scalar,
         ty: Ty<'tcx>,
     ) -> &'tcx Self {
-        Self::from_const_value(tcx, ConstValue::from_primval(val), ty)
+        Self::from_const_value(tcx, ConstValue::from_scalar(val), ty)
     }
 
     #[inline]
     pub fn from_bits(
         tcx: TyCtxt<'_, '_, 'tcx>,
-        val: u128,
-        ty: Ty<'tcx>,
+        bits: u128,
+        ty: ParamEnvAnd<'tcx, Ty<'tcx>>,
     ) -> &'tcx Self {
-        Self::from_primval(tcx, PrimVal::Bytes(val), ty)
+        let ty = tcx.global_tcx().lift(&ty).unwrap();
+        let defined = tcx.global_tcx().layout_of(ty).unwrap_or_else(|e| {
+            panic!("could not compute layout for {:?}: {:?}", ty, e)
+        }).size.bits() as u8;
+        Self::from_scalar(tcx, Scalar::Bits { bits, defined }, ty.value)
```

Comment:
I have two potential suggestions:
* don't actually get the layout here, use `128` (we did get that many bits)
* truncate to the known bit count and assert that the result of the truncation is identical (i.e. there are no stray `1`s)

Reply:
The latter is also great because you can add a `from_bits_trunc` that doesn't assert, just silently truncates. Useful for signed values (or maybe there could be a `from_bits_signed` that takes an `i128` and asserts that sign-extending the result of the truncation results in the same original value).